File: Number_of_Islands.py
class Solution:
    def numIslands(self, grid):
        if not grid:
            return 0
        m, n = len(grid), len(grid[0])
        # gone = [[0 for i in range(n)] for j in range(m)]
        
        
        def dfs(grid, i, j):
            if i<0 or j<0 or i>=m or j>=n or grid[i][j] != '1':
                return 
            grid[i][j] = '#'
            dfs(grid, i+1, j)
            dfs(grid, i-1, j)
            dfs(grid, i, j+1)
            dfs(grid, i, j-1)
        

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' :
                    dfs(grid, i, j)
                    count += 1
        return count

File: Number_of_Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty()) return 0;
        int m = grid.size(), n = grid[0].size();
        int land_num = 0;
        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                if (grid[i][j] == '1'){
                    land_num++;
                    traverse(i, j, grid);
                }         
            }
        }
        return land_num;
    }
    
    void traverse(int m, int n, vector<vector<char>>& grid){
        if (m < 0 || n < 0 || m == grid.size() || n == grid[0].size() || grid[m][n] == '0')
            return;
        grid[m][n] = '0';
        traverse(m+1, n, grid);
        traverse(m, n+1, grid);
        traverse(m-1, n, grid);
        traverse(m, n-1, grid);
    }
};